### PR TITLE
Add betrayal scenario support for original campaign

### LINKED ISSUE #3491: Campaign misses two scenarios that allows hero to betray his Lord
image
image
Player should be able to "Choose his Lord" once again. :)

### LINKED ISSUE #4985: Wrong scenario order in "Voyage Home" campaign
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Campaign continues after finishing 3rd scenario. In the OG you complete campaign either by finishing 3 or 4 scenario.
image

### Save file

[Campaig wrong order.zip](https://github.com/ihhub/fheroes2/files/8007754/Campaig.wrong.order.zip)

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by ihhub**
This pull request fixes multiple issues related to campaigns:
- fix multiple warnings / code smells detected by static analysers
- fix logic of detecting last scenario of a campaign ( #4985 ), especially `isLastScenario` method. Also scenario sequence was incorrect
- add an ability of betraying a lord in the original campaign. Previously we were working only with scenario ID assuming that we can't change campaigns. In the proposed code we save scenario ID and campaign ID for each scenario. This allows us to track which scenarios are completed and which will lead to a different campaign
- add a popup confirmation window for scenario restart
- some code cleanup and optimization

close #3491
close #4985

### CODE REVIEW

**Comment by a1exsh** [comment]
@ihhub you beat me to it! :D

**Review by oleg-derevenetz — COMMENTED** [review]
Hi @ihhub no obvious issues caught my attention, except one minor note. Please take a look when it's convenient for you.

File: src/fheroes2/campaign/campaign_scenariodata.h [hunk 1]
```diff
@@ -50,6 +50,33 @@ namespace Campaign
         LOSE_ALL_SORCERESS_VILLAGES = 1
     };
 
+    struct ScenarioInfoId
+    {
+        ScenarioInfoId() = default;
+
+        // Make sure that we don't pass just one argument when 2 are required.
+        ScenarioInfoId( const int ) = delete;
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/campaign/campaign_scenariodata.h]: I am not sure why it was working but somehow during the changes 1 argument was acceptable as well. Now it is not. I removed it.

**Comment by ihhub** [comment]
Hi @Branikolog , could you please test these changes as the highest priority?

**Comment by a1exsh** [comment]
@ihhub cool, seems like it just works! :D

I gave it a try and I notice that the audio sequence of Roland addressing the player with "Come and fight on my side" is missing in the Archibald's campaign.

The Archibald addressing the player in Roland's campaign is present, but if the player hits a key early to skip intro he also skips the 2nd sequence.  This seems to be the OG problem, actually, i.e. in one campaign there are two separate sequences, but in the other — just one.

**Comment by Branikolog** [comment]
Hello, @ihhub 
I've noticed next thing:
In the betrayal scenario we don't have previously earned rewards. (they are visible at scenario selection window, but not applied after start)
OG:
image
fheroes2:
image
The next scenario this reward is still applied too.
OG has none:
image
fheroes2:
image

Also, I've noticed, that RESTART button doesn't perform confirmation dialog.
OG:
image

**Comment by ihhub** [comment]
Hi @a1exsh and @Branikolog , thanks for your findings! I will fix them soon.

**Comment by oleg-derevenetz** [comment]
@ihhub how about we hold off on the new release for a week to include this (well-tested) PR in it?

**Comment by ihhub** [comment]
> @ihhub how about we hold off on the new release (0.9.12) for a week to include this (well-tested) PR in it?

I've marked it for the next release :)

**Comment by oleg-derevenetz** [comment]
@ihhub

> I've marked it for the next 0.9.13 release :)

Yeah, that's why I asking :) This is a good change and it's not fair that people will only get it in a month :)

**Comment by ihhub** [comment]
> @ihhub
> 
> > I've marked it for the next 0.9.13 release :)
> 
> Yeah, that's why I'm asking :) This is a good change and it's not fair that people will only get it in a month :)

Let me try then to fix the rest of issues fast. I'll do it today and we'll test again.

**Comment by ihhub** [comment]
Hi @a1exsh and @Branikolog , I addressed all your comments. Could you please take a look again?

@Branikolog , also please verify that other campaigns work properly and old saves are still in working state.

**Review by oleg-derevenetz — CHANGES_REQUESTED** [review]
Hi @ihhub I just noticed a few TYPOS, nothing serious.

**Comment by a1exsh** [comment]
@ihhub the missing video sequence is in place now.  Great piece of work!

**Comment by ihhub** [comment]
> Hi @ihhub I just noticed a few typos, nothing serious.

Thank you for the review! I've fixed the typos.

**Pull request merged**